SI-9750 scala.util.Properties.isJavaAtLeast works with JDK9

som-snytt opened the pull request:
The utility method compares javaSpecVersion, which has the
form "1.8" previously and "9" going forward.

The method accepts "1.n" for n < 9. More correctly, the string
argument should be a single number.

Supports JEP-223.

Supersedes https://github.com/scala/scala/pull/5270

Review by @golem131

This supposes golem was annoyed by our preciousness.

Commit:
SI-9750 scala.util.Properties.isJavaAtLeast works with JDK9

The utility method compares javaSpecVersion, which has the
form "1.8" previously and "9" going forward.

The method accepts "1.n" for n < 9. More correctly, the string
argument should be a single number.

Supports JEP-223.

som-snytt commented on src/library/scala/util/Properties.scala:
```diff
@@ -168,27 +168,31 @@ private[scala] trait PropertiesTrait {
+    def versionOf(s: String): Int = s.indexOf('.') match {
+      case 1 if s.charAt(0) == '1' =>
+        val v = versionOf(s.substring(2))
+        if (v < 9) v else -1
```
Probably didn't intend to accept `1.1.8`.

som-snytt commented on src/library/scala/util/Properties.scala:
```diff
@@ -168,27 +168,31 @@ private[scala] trait PropertiesTrait {
-    val (s, _s) = parts(javaSpecVersion)
-    s.toInt >= v.toInt && _s.toInt >= _v.toInt
+    val v = versionOf(version)
+    if (v < 0) throw new NumberFormatException(s"Not a version: $version")
```
Is that the clearest message? `Not a version: 1.9`. Use `assertThrows` to verify the message in a test.

som-snytt committed:
SI-9750 Tweak tests for what is a number

Leaves the error string as is, but adds test to show how it looks.
Java calls it a version number. `Not a version: 1.9`.

Don't strip `1.` prefix recursively. (That was Snytt's fault.)

som-snytt committed:
SI-9750 isJavaAtLeast(Int)

A good opportunity to simplify the API. Versions are strings,
but a spec version is just a number.

varming commented on src/library/scala/util/Properties.scala:
```diff
@@ -168,27 +168,31 @@ private[scala] trait PropertiesTrait {
-      val i = x.indexOf('.')
-      if (i < 0) throw new NumberFormatException("Not a version: " + x)
-      (x.substring(0, i), x.substring(i+1, x.length))
+    def versionOf(s: String): Int = s.indexOf('.') match {
```
[JEP223](http://openjdk.java.net/jeps/223) specify that the system property java.specification.version is a $VNUM and $VNUMs satisfy the regular expression: `^[1-9][0-9]*(((\.0)*\.[1-9][0-9]*)*)*$`. This code doesn't look ready to pick the major version from all strings satisfying that regular expression.

som-snytt replied:
@varming You're right, it does say $VNUM. I was fooled by the table of examples where the spec is $MAJOR? Is the spec $MAJOR in practice? I think it can't hurt to accept leading "1." and then ignore suffix of ".anything". The doc should say the arg must be the $MAJOR, if that is the desired behavior. Certainly previous usage has been just to distinguish major versions of the platform.

som-snytt replied:
Actually, re-reading what they intend by $MINOR, which includes revision of API, it's possible this API should support $MAJOR.$MINOR. I know I care if we're on java 8 or 9, but will I care if it is 9.1 or 9.2? assuming the spec string is 9.1. Alternatively, maybe supply `isJavaMajorSpec` for platforms that don't want to use regex for this test; otherwise just use regex like normal people.

varming replied:
Dear @som-snytt,

I don't know if there will be a time where $minor, $security, and later parts of the $VNUM will be important, but I would expect the code to work with all $VNUMs. In most settings you probably want to support $major, $minor, and $security. Having special functions to check that the major is at least greater than some integer seems like a good idea.

Right now the code handles backwards compatibility with <= JDK8, but it cannot handle JDK10. It would be nice if it handled backwards compatibility separately from general $VNUMs and let backwards compatibility take precedence. It doesn't look like you need full regular expressions to parse a $VNUM, simply take all characters up to the next dot (or the end of the string if there is no more dots) and parse them as an integer, and repeat this until you can decide the version question.

som-snytt committed:
SI-9750 Spec check major.minor.security

Don't assume spec is just major, but allow arbitrary version
number for both spec value and user value to check.

Only the first three dot-separated fields are considered,
after skipping optional leading value "1" in legacy format.

Minimal validity checks of user arg are applied. Leading three
fields, if present, must be number values, but subsequent
fields are ignored.

Note that a version number is not a version string, which
optionally includes pre and build info, `9-ea+109`.

som-snytt commented:
By coincidence, there's a performance regression on start-up in JDK-9 that has to do with loading regex classes, fixed by checking if the version is just a major number.  [The call for review.](http://mail.openjdk.java.net/pipermail/core-libs-dev/2016-July/042538.html)

After the nuclear apocalypse, the only things left will be cockroaches and programs that do `s.indexOf('.')`.

retronym commented:
Looks like we can actually remove the only use of this method from our own code. Would you mind doing that too? We know statically that `isJavaAtLeast("1.7")` is true.

```
val suppressable = isJavaAtLeast("1.7")
...
      if (suppressable) {
        import scala.language.reflectiveCalls
        type Suppressing = { def getSuppressed(): Array[Throwable] }
        for (s <- e.asInstanceOf[Suppressing].getSuppressed) print(s, Suppressed, frames, indents + 1)
      }
```

som-snytt committed:
SI-9750 Remove isJavaAtLeast from util.StackTracing

Formatting suppressed exceptions required reflection for platform
compatibility. No longer, since Java 8 is assumed. Minor tidying.

som-snytt commented:
Sure thing, it gave me something to do while babysitting my second Windows 10 upgrade this week, with a few hours to spare.

I only wish that guy would comment his code. Frankly, I never want to touch his code ever again. I don't see how anyone puts up with it.

Technically, `isJavaAtLeast(Int)` still calls `isJavaAtLeast(String)`.

SethTisue commented:
shouldn't this target 2.11.x?

som-snytt commented:
@SethTisue Too risky!

retronym commented:
LGTM. I think we can get target Java 9 support to the 2.12.1+, and then backport to 2.11.x once some problems are shaken out.
